syntax = "proto3";

package p2p2p;

option go_package = "go/types/message;message";

// used to wrap encrypted messages
message MessageWrapper {
  bytes message = 1;// encrypted
  bytes signature = 2;
  bytes nonce = 3;
}

message Message {
  MessageType type = 1;
  oneof payload {
    FileDownloadRequeset file_download_request = 2;
    FileList file_list = 3;
    File file = 4;
  }
}

// first thing sent by the client after secure connection is established
enum MessageType {
  FILE_DOWNLOAD_REQUEST = 0;
  FILE_UPLOAD_REQUEST = 1;
  FILE_LIST_REQUEST = 2;
  FILE = 3;
  FILE_LIST = 4;
  DECLINE = 5;
  CONFIRMATION = 6;
}

message Decline { // used to decline a request to perform an action
  string reason = 1;
}

message Confirmation { // used to confirm action was successful
  string message = 1;
}

// Protocol messages
message FileDownloadRequeset {
    string file_name = 1;
}

message FileUploadRequest {
    string file_name = 1;
}

message FileListRequest {

}

message KeyMigrationRequest {
}

message FileList {
    repeated FileMetadata files = 1;
}

message FileMetadata {
    string name = 2;
}

message File {
  string name = 1;
  bytes data = 2;
}